module/zfs/zfs_log.c

```diff
@@ -524,6 +525,113 @@ zfs_log_rename(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype, znode_t *sdzp,
+ *
+ * Ideally, these operations should be represented as new ZFS Intent Log
+ * txtypes, which would mandate a new ZFS feature flag due to the on-disk
+ * format change.  One would then use spa_feature_incr/decr functions to
+ * indicate that the on-disk log contains these new txtypes.  However, these
+ * functions are only supposed to be called from the txg syncing context.
+ *
+ * This means that we would need to force out an in-progress txg to disk and
+ * start a new one before writing any ZIL records.  This would ensure that
+ * previous versions of ZFS which do not support these log txtypes would
+ * never encounter them during ZIL replay.  Doing this would hurt performance.
+ *
+ * Alternatively, we could just activate the feature on a pool when these
+ * renameat2 flags get first used and leave it at that.  This would render
+ * the pool read-only importable on implementations without the new feature
+ * flag, even when no new txtypes were present on-disk.  This could be almost
+ * all of the time, so it'd be a shame to render the pool read-only on
+ * non-Linux platforms.
+ *
+ * Instead, we choose to rely on the fact that the ZIL is replayed in single-
+ * threaded mode before the dataset is mounted.  This means we can represent
+ * the otherwise atomic operations as a series of plain good old txtypes
+ * known to all current OpenZFS implementations. To do that, we use the
+ * following functions (at least until more platforms implement renameat2).
+ *
+ *     zfs_log_rename_exchange
+ *     zfs_log_rename_whiteout
+ */
+
+void
+zfs_log_rename_exchange(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype,
+    znode_t *sdzp, const char *sname, znode_t *tdzp, const char *dname,
+    znode_t *szp)
+{
+#ifndef __FreeBSD__
+	zfs_dirlock_t *tmpdl;
+#endif
+	znode_t *tmpzp = NULL;
+	char *tmpname;
+	int retries = 0;
+	int pos = 0;
+	int error;
+
+	tmpname = kmem_alloc(MAXPATHLEN, KM_SLEEP);
+
+	/*
+	 * To represent atomic rename with old non-atomic operations, we need
+	 * a temporary new name; so we try picking a name until we succeed,
+	 * then we get a dirent lock for that temp name until the final itx
+	 * gets queued
+	 */
+retry:
+	retries++;
+	pos = snprintf(tmpname, MAXPATHLEN, "%s.zfs_renameat2_emul_", dname);
+
+	for (int i = 0; i < 16; i++) {
+		int r = 0xFF;
+		random_get_pseudo_bytes((void *)&r, 1);
+		pos += snprintf(tmpname+pos, MAXPATHLEN-pos, "%02x", r);
+	}
+
+#ifdef __FreeBSD__
+	error = zfs_dirent_lookup(tdzp, tmpname, &tmpzp, ZNEW);
+#else
+	error = zfs_dirent_lock(&tmpdl, tdzp, tmpname, &tmpzp, ZNEW, NULL, NULL);
+#endif
+
+	VERIFY3U(retries, <, 10);
+	if (error)
+		goto retry;
+
+	/* dst -> tmp */
+	zfs_log_rename(zilog, tx, txtype, tdzp, dname, tdzp, tmpname, szp);
+	/* src -> dst */
+	zfs_log_rename(zilog, tx, txtype, sdzp, sname, tdzp, dname, szp);
+	/* tmp -> src */
+	zfs_log_rename(zilog, tx, txtype, tdzp, tmpname, sdzp, sname, szp);
+
+#ifdef __FreeBSD__
+	VOP_UNLOCK1(tmpzp);
```

Comment:
@behlendorf I'm not entirely sure what is the right thing to do here (FreeBSD doesn't have `zfs_dirent_lock` but it has `zfs_dirent_lookup` which appears to be similar). However, the documentation in `module/os/freebsd/zfs/zfs_vnops_os.c` says you can "unlock" the `tmpzp` using `zfs_dirent_unlock` -- which doesn't exist in FreeBSD:

```c
/* [snip]
 * In general, this is how things should be ordered in each vnode op:
 *
 *	ZFS_ENTER(zfsvfs);		// exit if unmounted
 * top:
 *	zfs_dirent_lookup(&dl, ...)	// lock directory entry (may VN_HOLD())
 *	rw_enter(...);			// grab any other locks you need
 *	tx = dmu_tx_create(...);	// get DMU tx
 *	dmu_tx_hold_*();		// hold each object you might modify
 *	error = dmu_tx_assign(tx, (waited ? TXG_NOTHROTTLE : 0) | TXG_NOWAIT);
 *	if (error) {
 *		rw_exit(...);		// drop locks
 *		zfs_dirent_unlock(dl);	// unlock directory entry
 *		VN_RELE(...);		// release held vnodes
 *		if (error == ERESTART) {
 *			waited = B_TRUE;
 *			dmu_tx_wait(tx);
 *			dmu_tx_abort(tx);
 *			goto top;
 *		}
 *		dmu_tx_abort(tx);	// abort DMU tx
 *		ZFS_EXIT(zfsvfs);	// finished in zfs
 *		return (error);		// really out of space
 *	}
 *	error = do_real_work();		// do whatever this VOP does
 *	if (error == 0)
 *		zfs_log_*(...);		// on success, make ZIL entry
 *	dmu_tx_commit(tx);		// commit DMU tx -- error or not
 *	rw_exit(...);			// drop locks
 *	zfs_dirent_unlock(dl);		// unlock directory entry
 *	VN_RELE(...);			// release held vnodes
 *	zil_commit(zilog, foid);	// synchronous when necessary
 *	ZFS_EXIT(zfsvfs);		// finished in zfs
 *	return (error);			// done, report error
 */
```

Reply:
Ah, `zfs_dirent_lookup` (as well as this comment) comes from Solaris it seems -- so it probably never applied to FreeBSD and was just copied along in the merge. Yeah this amount of spelunking is a bit out of my depth, so a simple answer to this question would be much appreciated. :sweat_smile:

Reply:
@asomers would you have some idea what we'll need for FreeBSD here? Or if not maybe you can suggest someone else who might.

Reply:
@freqlabs here's the commit that replaced zfs_dirent_lock, with the explanation why.  Is that what you're looking for?
https://github.com/freebsd/freebsd-src/commit/f79bc172334049435a3aa983e11e5ac494169d53

Reply:
Looking at that commit, it seems that we don't need to do any unlocking after `zfs_dirent_lookup` (all of the `zfs_dirent_unlock` calls were deleted and not replaced with anything else)?

Reply:
Yeah it looks like FreeBSD's version of lockdep doesn't like me doing `zfs_dirent_lookup` without an unlock step...

Reply:
As an aside it doesn't actually matter what we do for FreeBSD here because the code is never run on FreeBSD (at least not until FreeBSD supports `renameat2(2)` flags).